stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +112,331 @@ internal func _tryNormalize(
+  do {
+    let (read, filled) = fastFill(rebasedSourceBuffer, outputBuffer)
+      if filled > 0 {
+        let nextIndex = readIndex.encoded(offsetBy: read)
+        _internalInvariant(sourceBuffer.isOnUnicodeScalarBoundary(nextIndex.encodedOffset))
+        
+        return NormalizationResult(
+          amountFilled: filled, nextReadPosition: nextIndex, reallocatedBuffers: false
+        )
+      }
+  }
+  guard let (read, filled) = transcodeSegmentToUTF16(rebasedSourceBuffer, into: icuInputBuffer)
+  else {
+    return _fastReallocateBuffers(
+      readIndex: readIndex,
+      sourceBuffer: sourceBuffer,
+      outputBuffer: &outputBuffer,
+      icuInputBuffer: &icuInputBuffer,
+      icuOutputBuffer: &icuOutputBuffer
+    )
+  }
+  
+  let nextIndex = readIndex.encoded(offsetBy: read)
+  _internalInvariant(sourceBuffer.isOnUnicodeScalarBoundary(nextIndex.encodedOffset))
+  
+  let rebasedICUInputBuffer = UnsafeBufferPointer(rebasing: icuInputBuffer[..<filled])
+  guard let normalized = _tryNormalize(rebasedICUInputBuffer, into: icuOutputBuffer) else {
+    return _fastReallocateBuffers(
+      readIndex: readIndex,
+      sourceBuffer: sourceBuffer,
+      outputBuffer: &outputBuffer,
+      icuInputBuffer: &icuInputBuffer,
+      icuOutputBuffer: &icuOutputBuffer
+    )
+  }
+  
+  guard let (_, transcoded) = transcodeSegmentToUTF8(
+    UnsafeBufferPointer<UInt16>(rebasing: icuOutputBuffer[..<normalized]),
+    into: outputBuffer
+  ) else {
+    return _fastReallocateBuffers(
+      readIndex: readIndex,
+      sourceBuffer: sourceBuffer,
+      outputBuffer: &outputBuffer,
+      icuInputBuffer: &icuInputBuffer,
+      icuOutputBuffer: &icuOutputBuffer
+    )
+  }
+  return NormalizationResult(
+    amountFilled: transcoded, nextReadPosition: nextIndex, reallocatedBuffers: false
+  )
+}
+
+internal func _foreignNormalize(
+  readIndex: String.Index,
+  endIndex: String.Index,
+  guts: _StringGuts,
+  outputBuffer: inout UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: inout UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: inout UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  guard let (read, filled) = foreignFill(
+    readIndex: readIndex, endIndex: endIndex, guts, into: icuInputBuffer
+  ) else {
+    return _foreignReallocateBuffers(
+      readIndex: readIndex,
+      endIndex: endIndex,
+      guts: guts,
+      outputBuffer: &outputBuffer,
+      icuInputBuffer: &icuInputBuffer,
+      icuOutputBuffer: &icuOutputBuffer
+    )
+  }
+  
+  let nextIndex = readIndex.encoded(offsetBy: read)
+  _internalInvariant(guts.isOnUnicodeScalarBoundary(nextIndex))
+  
+  let rebasedICUInputBuffer = UnsafeBufferPointer(rebasing: icuInputBuffer[..<filled])
+  guard let normalized = _tryNormalize(rebasedICUInputBuffer, into: icuOutputBuffer) else {
+    return _foreignReallocateBuffers(
+      readIndex: readIndex,
+      endIndex: endIndex,
+      guts: guts,
```

Comment:
If this isn't a recursive call, then we don't need to pass guts vs the UBP. We just want to know the code unit count.